src/binary-reader.cc

```diff
@@ -2829,13 +2830,13 @@ Result BinaryReader::ReadCodeSection(Offset section_size) {
 
 Result BinaryReader::ReadDataSection(Offset section_size) {
   CALLBACK(BeginDataSection, section_size);
-  Index num_data_segments;
-  CHECK_RESULT(ReadCount(&num_data_segments, "data segment count"));
-  CALLBACK(OnDataSegmentCount, num_data_segments);
+  CHECK_RESULT(ReadCount(&num_data_segments_, "data segment count"));
+  CALLBACK(OnDataSegmentCount, num_data_segments_);
   // If the DataCount section is not present, then data_count_ will be invalid.
-  ERROR_UNLESS(data_count_ == kInvalidIndex || data_count_ == num_data_segments,
-               "data segment count does not equal count in DataCount section");
-  for (Index i = 0; i < num_data_segments; ++i) {
+  ERROR_UNLESS(
+      data_count_ == kInvalidIndex || data_count_ == num_data_segments_,
+      "data segment count does not equal count in DataCount section");
```

Comment:
I think this check can not be removed, since the module level check will always catch this error.

Reply:
do we want to keep parsing when we encounter a malformed module?

Reply:
@sbc100 ping?

Reply:
> we think doing this check early is important

Can you explain why?  It seems redundant to me.

Reply:
Whoops, missed this. Uh, we think, because the module is malformed, it's better to exit early if we can. Tho the more important thing is that a malformed module shouldn't crash the parser, we guess.

Reply:
I agree the parse should not crash.   Are you saying that if you remove these 3 lines the parser will crash?   Where/why?

Reply:
> , because the module is malformed, it's better to exit early if we can. 

I think delaying the error would be fine in this case, it it means we can avoid the duplication.

Reply:
Tweaked and rebased, let's see what CI has to say.

Reply:
well.

based on the changes to the error messages, we prefer the duplicated checks.